src/databricks/labs/lsql/dashboards.py

```diff
@@ -728,6 +767,120 @@ def _get_query_widget_spec(fields: list[Field], *, frame: WidgetFrameSpec | None
+            show_title=len(self.metadata.title) > 0,
+            description=self.metadata.description,
+            show_description=len(self.metadata.description) > 0,
+        )
+        control_encoding_map = ControlEncodingMap(control_encodings)
+        spec = MultiSelectSpec(encodings=control_encoding_map, frame=frame)
+        widget = Widget(name=f"{self.metadata.id}_widget", queries=queries, spec=spec)
+        return widget
+
+    def _create_date_range_picker_widget(self, dataset_columns: set[tuple[str, str]]) -> Widget:
+        control_encodings: list[ControlEncoding] = []
+        queries = []
+        for idx, (column, dataset_name) in enumerate(dataset_columns):
+            filter_fields = [
+                Field(name=column, expression=f"`{column}`"),
+                Field(name=f"{column}_associativity", expression="COUNT_IF(`associative_filter_predicate_group`)"),
+            ]
+            filter_query = Query(dataset_name=dataset_name, fields=filter_fields, disaggregated=False)
+            filter_named_query = NamedQuery(name=f"date_range_picker_{column}_{idx}", query=filter_query)
+            queries.append(filter_named_query)
+            control_encodings.append(ControlFieldEncoding(column, filter_named_query.name, display_name=column))
+
+        frame = WidgetFrameSpec(
+            title=self.metadata.title,
+            show_title=len(self.metadata.title) > 0,
+            description=self.metadata.description,
+            show_description=len(self.metadata.description) > 0,
+        )
+        control_encoding_map = ControlEncodingMap(control_encodings)
+        spec = DateRangePickerSpec(encodings=control_encoding_map, frame=frame)
+        widget = Widget(name=f"{self.metadata.id}_widget", queries=queries, spec=spec)
+        return widget
+
+    def _get_dataset_columns(self, datasets: list[Dataset]) -> set[tuple[str, str]]:
+        """Get the filter column and dataset name pairs."""
+        dataset_columns = set()
+        for dataset in datasets:
+            for field in self._find_filter_fields(dataset.query):
+                dataset_columns.add((field.name, dataset.name))
+        return dataset_columns
+
+    def _find_filter_fields(self, query: str) -> list[Field]:
+        """Find the fields in a query matching the filter names.
+        The fields are the projections in the query's top level SELECT.
+        """
+        try:
+            abstract_syntax_tree = sqlglot.parse_one(query, dialect=_SQL_DIALECT)
+        except sqlglot.ParseError as e:
+            logger.warning(f"Error while parsing {query}: {e}")
+            return []
+        filters = {name.lower() for name in self.metadata.filters}
+        filter_fields = []
+        for projection in abstract_syntax_tree.find_all(sqlglot.exp.Select):
+            if projection.depth > 0:
+                continue
+            for named_select in projection.named_selects:
+                if named_select.lower() in filters:
```

Comment:
nit: reduce nesting, rewrite with `if ... not in ...: continue`. be consistent with lines 875-876.

Reply:
Done